src/language/completionItemProvider.ts

```diff
@@ -0,0 +1,51 @@
+import {
+  ColumnDefinition,
+  getStandardSQLCompletionProvider,
+  LanguageCompletionProvider,
+  LinkedToken,
+  TableDefinition,
+  TableIdentifier,
+} from '@grafana/experimental';
+import { MACROS } from './macros';
+
+interface CompletionProviderGetterArgs {
+  getDatabases: React.MutableRefObject<() => Promise<TableDefinition[]>>;
+  getTables: React.MutableRefObject<(d?: string) => Promise<TableDefinition[]>>;
+  getColumns: React.MutableRefObject<(database?: string, table?: string) => Promise<ColumnDefinition[]>>;
+}
+
+export const getTimestreamCompletionProvider: (args: CompletionProviderGetterArgs) => LanguageCompletionProvider = ({
+  getDatabases,
+  getTables,
+  getColumns,
+}) => (monaco, language) => {
+  return {
+    // get standard SQL completion provider which will resolve functions and macros
+    ...(language && getStandardSQLCompletionProvider(monaco, language)),
+    triggerCharacters: ['.', ' ', '$', ',', '(', "'"],
+    schemas: {
+      resolve: async () => getDatabases.current(),
+    },
+    tables: {
+      resolve: async (t: TableIdentifier) => {
+        return await getTables.current(t?.schema);
+      },
+      parseName: (token: LinkedToken) => {
+        let tablePath = token?.value ?? '';
+
+        const parts = tablePath.split('.');
+        if (parts.length === 1) {
+          return { schema: parts[0] || undefined };
+        } else if (parts.length === 2) {
+          return { schema: parts[0] || undefined, table: parts[1] || undefined };
+        }
```

Comment:
SUPER NITPICKY (feel free to ignore) just feels like this could be made easier somehow idk
```
const tablePath = token?.value ?? '';
const parts = tablePath.split('.');
const schema = parts.length >= 1 && parts[0];
const table = parts.length === 2 && parts[1]
if (parts.length > 1) {
    return { schema, table }
}
return null
```